storagenode/piecestore/endpoint.go

```diff
@@ -181,6 +182,28 @@ func (endpoint *Endpoint) DeletePieces(
 	}, nil
 }
 
+func (endpoint *Endpoint) flagUnusualSlowUpload(averageUploadRate float64, connectionCount int64) bool {
+	// congestionThreshold defines the proportion of MaxConcurrentRequests
+	// that must be reached before the storage node will begin dropping slow
+	// upload connections to preserve its own bandwidth.
+	const congestionThreshold = 0.8
+
+	if averageUploadRate < float64(endpoint.config.MinUploadSpeed) {
```

Comment:
Note, this would also need to handle recovering from a congestion overload.

e.g. when you get a torrent of uploads then after recovery, there could still be some connections that haven't picked up the speed.

I'm not sure what the best way to handle this whole logic would be.

Maybe entirely a separate struct for tracking the upload speed? I haven't thought this completely through, but:

```
type uploadSpeedEstimation struct {
	startTime               time.Time
	congested               bool
	recoveredFromCongestion time.Time
	recoveredSize           int64 // how much had been transferred by the time of recovery
}
```

Reply:
Would you please explain a bit further ?
 
If I get what you mean, we may want to take into account case when the whole network is congested and the connection become globally slow ?

What do you mean by "recovery" in this case ? Or a recovery from a  connection dropped ?

Reply:
For example, currently this slow detection switches off when there are a lot of incoming connections. Once we recover from that "congested" state the left over connections would still have uploaded only very few bytes, hence they would be reaped by the slow upload check.

To give an example (with max concurrency 100):

1. 100 uploads are started, which progress very slowly
2. isSlowUpload check won't cancel the uploads because it's considered congested
3. lets say this congestion lasts for 30 seconds, then 30 uploads are canceled
4. in those 30 seconds all other 70 uploads haven't made progress due to congestion
5. all the other 70 uploads get cancelled, because of the min upload limit

Reply:
> For example, currently this slow detection switches off when there are a lot of incoming connections. Once we recover from that "congested" state the left over connections would still have uploaded only very few bytes, hence they would be reaped by the slow upload check.
> 
> To give an example (with max concurrency 100):
> 
> 1. 100 uploads are started, which progress very slowly
> 2. isSlowUpload check won't cancel the uploads because it's considered congested
> 3. lets say this congestion lasts for 30 seconds, then 30 uploads are canceled
> 4. in those 30 seconds all other 70 uploads haven't made progress due to congestion
> 5. all the other 70 uploads get cancelled, because of the min upload limit

In 

```golang
func (endpoint *Endpoint) isSlowUpload(currentUploadSize float64, connectionDuration time.Duration) bool {
	// Skip flagging unusually slow connections in the first moments of uploading.
	if connectionDuration < endpoint.config.MinUploadSpeedGraceDuration || connectionDuration == 0 {
		return false
	}

	// Latest number of alive connection
	connectionCount := atomic.AddInt32(&endpoint.liveRequests, 1)

```

The `connectionCount` is actually refreshed at every upload chunk/iteration.

Thus for such use case, after a certain number of connections are closed , the `connectionCount` would decrease and thus even if the current connections may upload very slowly, it will not be kicked out due to the fact that there are available slots for connections.

Is that correct or I am missing something ?

Reply:
> it will not be kicked out due to the fact that there are available slots for connections.

The issue is that on average the nodes ended up slow due to congestion even though the congestion is gone.

For example:

1. upload stalls for 30s due to congestion
2. upload manages then to upload 1MB in 1s after congestion clears
3. the average speed with current measurement would be `258kb/s`.

If there's a ramp-up to get to full speed, then that could be even smaller.

Reply:
Even in such a case, the stalled connection would not be kicked out, because although average speed is 258Kb/s, well below the limit, but the number of the alive connection would not be above 80% (due to the fact that some of the connections are dropped), so the remaining connections would remain opened and having enough times to ramp-up.

Is that correct ?

Reply:
The number of alive connections is updated on every iteration, hence it'll be below the threshold.

For simplicity, lets say all other 99 connections are cancelled and only one connection remains. Now lets say the check happens immediately after all the connections are cancelled and the leftover connection doesn't manage to send any data within 0.001s. At 30.001s mark the average speed would be 0kb/s.

So, after coming out of congested status, a single connection might not have sufficiently sent data to be above the min upload speed threshold.

Reply:
Agreed on that.

However, I think the original purpose of the issue mentioned https://github.com/storj/storj/issues/614 is to flag out and drop explicit malicious behavior. E.g: clients that intentionally slow down its upload to harm the network.

In my opinion, the mentioned above congestion problem makes sense completely, but is another set of problem that links to congestion/connection handling of the storage node. In order to do so, storage node shall have a tighter control on the available connections, i.e: better global connection handling.

The scenario this change can be useful:

- Only 20% capacity is available
- A set of clients initiate a slow loris attack

Behavior expected: Those set of clients shall be terminated. 
If there are other clients that intentionally slow down the upload, it shall be terminated as well.
For other clients that are considered to be "non-malicious", it shall not be affected.

What cannot be solved (by current changes):

- Congestion problem in the network that result it slow clients being kicked out.
- An attack is initiated thus results in global slow down of connection upload 

It seems that we would have to maintain a sufficiently complex state of each client if we want to achieve such congestion handling policy. Flag out individual client and kick them out is a stateless operation, but the proposed handling of congestion is a statefull action.

What do you think ?

Reply:
I was thinking, to simplify maintenance/development, how's about:

- Keeping this change limited to handling individual client, this means accepting false positive for the congestion case.
- Create another issue for better client connection handling in a stateful manner.

Make sense ?

Reply:
@egonelbre Any thoughts on this ?

Reply:
Sorry for not answering earlier, there were holidays yesterday and the day before in Estonia.

But, I would like to avoid false-positives as much as possible, since they cause more repair traffic on the network. We can assume that most uplinks actually want to upload data and rejecting their data is not great. I don't have great stats on how much that would end up cancelling -- it might be a negligible or it might be significant, hard to tell.

I need to think a bit, but I don't think the stateful would be that much more complex. I can come back with a a design in a few hours. Currently have to handle a few other things.

Reply:
Took a bit longer to get back to it, but I think this design should be clear enough and can be independently tested.

The estimator:

```
type speedEstimation struct {
	grace  time.Duration
	limit  memory.Size

	active      time.Duration
	lastChecked time.Time
}

func (estimate *speedEstimation) EnsureLimit(transferred memory.Size, congested bool, now time.Time) error {
	if estimate.lastChecked.IsZero() {
		estimate.lastChecked = now
		return nil
	}

	delta := now.Sub(estimate.lastChecked)
	estimate.lastChecked = now

	if congested {
		return nil
	}

	estimate.active += delta
	if estimate.active <= 0 || estimate.active <= estimate.grace {
		// not enough data
		return nil
	}

	bytesPerSec := transferred / memory.Size(estimate.active.Seconds())
	if bytesPerSec < estimate.limit {
		return Error.New("speed too low, current:%v < limit:%v", bytesPerSec, estimate.limit)
	}

	return nil
}
```

To hook it up:
```
[... snip ...]

speedEstimate := speedEstimation{
	grace: endpoint.config.MinUploadSpeedGraceDuration,
	limit: endpoint.config.MinUploadSpeed,
}

for {
	if err := speedEstimate.EnsureLimit(pieceWriter.Size(), endpoint.isCongested(), time.Now()); err != nil {
		return rpcstatus.Errorf(rpcstatus.Aborted, "upload rate fell below limit")
	}

[... snip ...]


func (endpoint *Endpoint) isCongested() bool {
	count := atomic.LoadInt32(&endpoint.liveRequests)
	return count > endpoint.config.MaxConcurrentRequests * 0.8
}
```

Reply:
Great !

The idea is to separate 2 concepts:
- Congested (global state)
- Slow speed (individual client and can be calculated at any moment in time)

To sum up, if I understand correctly, is that the slow speed check would be applied if the node is not in congested state.

Changes on the way . I'll try to implement it quickly with tests.

Reply:
@egonelbre thanks for the design. The necessary changes have been implemented.